optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/gizmo/GizmoSolutionClonerImplementor.java

```diff
@@ -0,0 +1,764 @@
+        methodCreator.returnValue(thisObj);
+    }
+
+    private static void createCloneSolution(ClassCreator classCreator, GizmoSolutionOrEntityDescriptor solutionInfo) {
+        Class<?> solutionClass = solutionInfo.getSolutionDescriptor().getSolutionClass();
+        MethodCreator methodCreator =
+                classCreator.getMethodCreator(MethodDescriptor.ofMethod(SolutionCloner.class,
+                        "cloneSolution",
+                        Object.class,
+                        Object.class));
+
+        ResultHandle thisObj = methodCreator.getMethodParam(0);
+
+        ResultHandle clone = methodCreator.invokeStaticMethod(
+                MethodDescriptor.ofMethod(
+                        GizmoSolutionClonerFactory.getGeneratedClassName(solutionInfo.getSolutionDescriptor()),
+                        "cloneSolutionRun", solutionClass, solutionClass, Map.class),
+                thisObj,
+                methodCreator.newInstance(MethodDescriptor.ofConstructor(IdentityHashMap.class)));
+        methodCreator.returnValue(clone);
+    }
+
+    private static void createCloneSolutionRun(ClassCreator classCreator, GizmoSolutionOrEntityDescriptor solutionInfo) {
+        Class<?> solutionClass = solutionInfo.getSolutionDescriptor().getSolutionClass();
+        MethodCreator methodCreator =
+                classCreator.getMethodCreator("cloneSolutionRun", solutionClass, solutionClass, Map.class);
+        methodCreator.setModifiers(Modifier.STATIC | Modifier.PRIVATE);
+
+        ResultHandle thisObj = methodCreator.getMethodParam(0);
+        ResultHandle createdCloneMap = methodCreator.getMethodParam(1);
+
+        ResultHandle maybeClone = methodCreator.invokeInterfaceMethod(
+                MethodDescriptor.ofMethod(Map.class, "get", Object.class, Object.class), createdCloneMap, thisObj);
+        BranchResult hasCloneBranchResult = methodCreator.ifNotNull(maybeClone);
+        BytecodeCreator hasCloneBranch = hasCloneBranchResult.trueBranch();
+        hasCloneBranch.returnValue(maybeClone);
+
+        BytecodeCreator noCloneBranch = hasCloneBranchResult.falseBranch();
+        ResultHandle clone = noCloneBranch.newInstance(MethodDescriptor.ofConstructor(solutionClass));
+
+        noCloneBranch.invokeInterfaceMethod(
+                MethodDescriptor.ofMethod(Map.class, "put", Object.class, Object.class, Object.class),
+                createdCloneMap, thisObj, clone);
+
+        for (GizmoMemberDescriptor shallowlyClonedField : solutionInfo.getShallowClonedMemberDescriptors()) {
+            writeShallowCloneInstructions(solutionInfo, noCloneBranch, shallowlyClonedField, thisObj, clone, createdCloneMap);
+        }
+
+        for (Field deeplyClonedField : solutionInfo.getDeepClonedFields()) {
+            GizmoMemberDescriptor gizmoMemberDescriptor = solutionInfo.getMemberDescriptorForField(deeplyClonedField);
+            final ResultHandle[] resultHandleHolder = new ResultHandle[1];
```

Comment:
If the point of this is to simply have an effectively final var to put in the lambdas, `AtomicReference<ResultHandle>` will probably make that point a bit clearer. Extraction of the handling code to a separate method would probably allow to get rid of either.

Reply:
So that what the mutable holder class in java called!

Why it is set up like this is to prevent errors from assuming the descriptor is a field or a method (or accidentally using the wrong one inside an if guard). That method was designed to have no return value (as the MemberAccessor code only use it for generating code, never for returning ResultHandles to the outside.

I could add a compute version of the method that takes two functions that returns ResultHandle (one accepts field descriptor, another accepts method descriptor), which removes the need to store the result in a holder.

Reply:
Done via adding Read/Write methods to GizmoMemberDescriptor